File: fundamentals_pipeline/51_streamlit_app/lib/screener.py
"""Screener data layer — one row per company from the published parquet artifacts.

Reuses :func:`lib.data.load_latest_data` (the same cached parquet loader the
detail page uses) — there is **no Databricks connection at runtime**. The wide
frame is built by pivoting the long-format `dashboard_metrics.parquet` to the
latest available fiscal year per ticker.

Universe flags (`is_favorite` / `in_sp500` / `in_r3000`) come from
`meta.tickers` and **Market Cap** comes from a `Market Cap` metric row — both
added by `50_publish/51__export_dashboard_data.py` (schema v3). Older artifacts
that predate v3 simply yield all-False flags and an empty Market Cap column; the
screener degrades gracefully until the next publish.
"""

from __future__ import annotations

import pandas as pd
import streamlit as st

from .data import load_latest_data

MARKET_CAP = "Market Cap"

# Columnas por defecto del screener (orden de aparición).
DEFAULT_COLUMNS = [
    "Market Cap", "P/E", "P/S", "Net Margin %", "ROE %", "Revenue YoY %", "EV/EBITDA",
]
# Métricas con slider de rango siempre visible.
RANGE_METRICS = ["Market Cap", "P/E", "Net Margin %", "ROE %", "Revenue YoY %"]
# Universo → columna de flag en el frame ("" = sin filtro).
UNIVERSE_FLAGS = {
    "Todos":        "",
    "S&P 500":      "in_sp500",
    "Russell 3000": "in_r3000",
    "Favoritos":    "is_favorite",
}
_FLAG_COLS = ("is_favorite", "in_sp500", "in_r3000")


def _as_bool(s: pd.Series) -> pd.Series:
    """Coerce a flag column to native bool, tolerant of legacy stringified values."""
    return s.map(lambda v: v is True or str(v).strip().lower() in ("true", "1")).astype(bool)


@st.cache_data(ttl=3600, show_spinner="Construyendo el screener…")
def build_screener_frame() -> tuple[pd.DataFrame, dict[str, str], list[str]]:
    """Return ``(wide, unit_map, metric_order)``.

    ``wide``: one row per ticker — columns ``ticker``, ``company``, the three
    universe flags, ``fiscal_year`` (the latest FY used) and one column per
    metric (pivoted from long format). ``unit_map``: metric → unit
    ('percent' / 'ratio' / 'usd') for display formatting. ``metric_order``:
    metrics ordered by `metrics_hierarchy.sort_order`, Market Cap first.
    """
    _data, metrics, meta = load_latest_data()

    m = metrics[metrics["period_type"] == "FY"].copy()
    # `metric` is a category dtype after _optimize_dtypes — cast to str so the
    # pivot doesn't materialise ghost columns for unused categories.
    m["metric"] = m["metric"].astype(str)

    # Quédate con el último fiscal_year disponible por ticker.
    latest = m.groupby("ticker", observed=True)["fiscal_year"].transform("max")
    m = m[m["fiscal_year"] == latest]

    # unit por métrica (para formateo) y orden por metrics_hierarchy.
    unit_map = (
        m.dropna(subset=["unit"]).drop_duplicates("metric")
        .assign(unit=lambda d: d["unit"].astype(str))
        .set_index("metric")["unit"].to_dict()
    )
    unit_map.setdefault(MARKET_CAP, "usd")

    order = (
        m.drop_duplicates("metric")
        .assign(so=lambda d: pd.to_numeric(d["sort_order"], errors="coerce"))
        .sort_values(["so", "metric"], na_position="last")["metric"].tolist()
    )
    metric_order = [MARKET_CAP] + [x for x in order if x != MARKET_CAP]

    # long → wide (en pandas; el SQL se mantiene simple).
    wide = m.pivot_table(index="ticker", columns="metric", values="value", aggfunc="first")
    wide.columns.name = None
    fy = m.groupby("ticker", observed=True)["fiscal_year"].max().rename("fiscal_year")
    wide = wide.join(fy).reset_index()

    # company + flags de universo desde meta.
    info = pd.DataFrame(meta.get("tickers", []))
    if "ticker" in info.columns:
        keep = ["ticker"] + [c for c in ("company", *_FLAG_COLS) if c in info.columns]
        wide = wide.merge(info[keep], on="ticker", how="left")
    if "company" not in wide.columns:
        wide["company"] = wide["ticker"]
    wide["company"] = wide["company"].fillna(wide["ticker"])
    for c in _FLAG_COLS:
        wide[c] = _as_bool(wide[c]) if c in wide.columns else False

    # Market Cap puede no existir aún (Release anterior a schema v3).
    if MARKET_CAP not in wide.columns:
        wide[MARKET_CAP] = pd.NA

    metric_order = [x for x in metric_order if x in wide.columns]
    return wide, unit_map, metric_order


def universe_mask(df: pd.DataFrame, universe: str) -> pd.Series:
    col = UNIVERSE_FLAGS.get(universe, "")
    if not col or col not in df.columns:
        return pd.Series(True, index=df.index)
    return df[col].astype(bool)


def search_mask(df: pd.DataFrame, query: str) -> pd.Series:
    q = (query or "").strip().lower()
    if not q:
        return pd.Series(True, index=df.index)
    return (
        df["ticker"].astype(str).str.lower().str.contains(q, na=False)
        | df["company"].astype(str).str.lower().str.contains(q, na=False)
    )


def range_bounds(series: pd.Series) -> tuple[float, float] | None:
    """Slider bounds clamped to the 1st–99th percentile (robust to outliers)."""
    s = pd.to_numeric(series, errors="coerce").dropna()
    if s.empty:
        return None
    lo, hi = float(s.quantile(0.01)), float(s.quantile(0.99))
    if lo == hi:
        hi = lo + 1.0
    return lo, hi


def range_mask(
    series: pd.Series, sel_lo: float, sel_hi: float, bound_lo: float, bound_hi: float
) -> pd.Series:
    """Mask for a range slider.

    A bound left at its slider extreme is treated as open-ended, so companies
    with values beyond the clamped percentile bounds are NOT filtered out at the
    default (full-span) position. NaN never hides a company (``| isna``).
    """
    s = pd.to_numeric(series, errors="coerce")
    m = pd.Series(True, index=s.index)
    if sel_lo > bound_lo:
        m &= s >= sel_lo
    if sel_hi < bound_hi:
        m &= s <= sel_hi
    return m | s.isna()

File: fundamentals_pipeline/60__frontends/62__web/src/lib/queries/screener.ts
// Screener data frame — the web equivalent of build_screener_frame() in
// 61__streamlit/lib/screener.py. One row per ticker, with the latest available fiscal-year
// value of every metric.
//
// Strategy: a single DuckDB query reduces the long metrics parquet to the latest-FY value
// per (ticker, metric), which we pivot to wide in JS (≈2,500 tickers — trivial). Sector,
// industry, company name and the universe flags come from meta.json `tickers` (NOT the
// parquet), joined here — matching the Streamlit `wide.merge(info, ...)`. Older artifacts
// missing a field degrade to its default (all-False flags, "Unknown" sector/industry).
//
// The whole frame is built once; column selection, filtering and sorting then happen
// in-memory in the table component (instant, no re-query) — same semantics as the cached
// Streamlit frame, and far simpler than a parametrised SQL round-trip per keystroke.

import { ARTIFACT_FILES, loadMeta, type TickerMeta } from "../data";
import { MARKET_CAP, UNKNOWN_INDUSTRY, UNKNOWN_SECTOR } from "../screener/constants";
import { queryRows } from "./runQuery";

export interface ScreenerRow {
  ticker: string;
  company: string;
  sector: string;
  industry: string;
  is_favorite: boolean;
  in_sp500: boolean;
  in_r3000: boolean;
  fiscal_year: number | null;
  /** metric name → latest-FY value (or null if absent). */
  metrics: Record<string, number | null>;
}

export interface ScreenerFrame {
  rows: ScreenerRow[];
  /** metric name → unit ("percent" | "ratio" | "usd"). */
  unitMap: Record<string, string>;
  /** metric names ordered by metrics_hierarchy.sort_order, Market Cap first. */
  metricOrder: string[];
}

/** Coerce a meta flag (bool, "true"/"1", or missing) to a native boolean — mirrors `_as_bool`. */
function asBool(v: unknown): boolean {
  return v === true || ["true", "1"].includes(String(v ?? "").trim().toLowerCase());
}

function cleanLabel(v: unknown, fallback: string): string {
  const s = String(v ?? "").trim();
  return s === "" ? fallback : s;
}

const METRICS = ARTIFACT_FILES.metrics;

export async function loadScreenerFrame(): Promise<ScreenerFrame> {
  // ── latest-FY value per (ticker, metric) ──────────────────────────────────────────────
  const factRows = await queryRows<{ ticker: string; metric: string; value: number; fiscal_year: number }>(
    `WITH fy AS (
       SELECT ticker, metric, fiscal_year, value
       FROM '${METRICS}'
       WHERE period_type = 'FY' AND value IS NOT NULL
     ),
     ranked AS (
       SELECT ticker, metric, value, fiscal_year,
              row_number() OVER (PARTITION BY ticker, metric ORDER BY fiscal_year DESC) AS rn
       FROM fy
     )
     SELECT ticker, metric, value, fiscal_year FROM ranked WHERE rn = 1`,
    ["metrics"],
  );

  // ── unit + sort order per metric (small: one row per metric) ──────────────────────────
  const metricMeta = await queryRows<{ metric: string; unit: string | null; sort_order: number | null }>(
    `SELECT metric, any_value(unit) AS unit, min(sort_order) AS sort_order
     FROM '${METRICS}' WHERE period_type = 'FY' GROUP BY metric`,
    ["metrics"],
  );

  const unitMap: Record<string, string> = {};
  for (const m of metricMeta) {
    if (m.unit != null) unitMap[m.metric] = String(m.unit);
  }
  if (!(MARKET_CAP in unitMap)) unitMap[MARKET_CAP] = "usd";

  // Order by sort_order (NaN/missing last), then name; Market Cap pinned first.
  const ordered = [...metricMeta].sort((a, b) => {
    const sa = a.sort_order ?? Number.POSITIVE_INFINITY;
    const sb = b.sort_order ?? Number.POSITIVE_INFINITY;
    return sa !== sb ? sa - sb : a.metric.localeCompare(b.metric);
  });
  const metricOrder = [MARKET_CAP, ...ordered.map((m) => m.metric).filter((m) => m !== MARKET_CAP)];

  // ── pivot long → wide in JS ───────────────────────────────────────────────────────────
  const byTicker = new Map<string, ScreenerRow>();
  for (const r of factRows) {
    let row = byTicker.get(r.ticker);
    if (!row) {
      row = {
        ticker: r.ticker,
        company: r.ticker,
        sector: UNKNOWN_SECTOR,
        industry: UNKNOWN_INDUSTRY,
        is_favorite: false,
        in_sp500: false,
        in_r3000: false,
        fiscal_year: null,
        metrics: {},
      };
      byTicker.set(r.ticker, row);
    }
    row.metrics[r.metric] = r.value;
    const fy = Number(r.fiscal_year);
    if (!Number.isNaN(fy) && (row.fiscal_year === null || fy > row.fiscal_year)) {
      row.fiscal_year = fy;
    }
  }

  // ── join meta.tickers (company, sector, industry, universe flags) ─────────────────────
  const meta = await loadMeta();
  const infoByTicker = new Map<string, TickerMeta>();
  for (const t of meta.tickers) {
    if (t?.ticker) infoByTicker.set(String(t.ticker), t);
  }
  for (const row of Array.from(byTicker.values())) {
    const info = infoByTicker.get(row.ticker);
    if (!info) continue;
    row.company = cleanLabel(info.company, row.ticker);
    row.sector = cleanLabel(info.sector, UNKNOWN_SECTOR);
    row.industry = cleanLabel(info.industry, UNKNOWN_INDUSTRY);
    row.is_favorite = asBool(info.is_favorite);
    row.in_sp500 = asBool(info.in_sp500);
    row.in_r3000 = asBool(info.in_r3000);
  }

  return { rows: Array.from(byTicker.values()), unitMap, metricOrder };
}
